### PR TITLE
Initial import from Gatekeeper + renaming

### PR DESCRIPTION
**Pull request by adrianludwin**
These two commit pass `go test ./...` successfully. I haven't tried importing it into HNC yet, but there are virtually no diffs between the HNC version and these files, so there should be minimal changes required.

### CODE REVIEW

File: pkg/rotator/rotator.go [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+	certsNotMounted chan struct{}
+}
+
+// Start starts the CertRotator runnable to rotate certs and ensure the certs are ready.
+func (cr *CertRotator) Start(stop <-chan (struct{})) error {
+	// explicitly rotate on the first round so that the certificate
+	// can be bootstrapped, otherwise manager exits before a cert can be written
+	crLog.Info("starting cert rotator controller")
+	defer crLog.Info("stopping cert rotator controller")
+	if err := cr.refreshCertIfNeeded(); err != nil {
```
Comment (ekitson) [inline pkg/rotator/rotator.go:90]: Just trying to make sure I understand the implications here: I think this is a blocking call; is that right? But the ensureCertsExist call is async? If so, these two calls seem at odds with each other?

File: pkg/rotator/rotator.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+)
+
+var crLog = logf.Log.WithName("cert-rotation")
+
+var vwhGVK = schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"}
+
+var _ manager.Runnable = &CertRotator{}
+
+// AddRotator adds the CertRotator and ReconcileVWH to the manager.
+func AddRotator(mgr manager.Manager, cr *CertRotator, vwhName string) error {
```
Comment (ekitson) [inline pkg/rotator/rotator.go:51]: Not sure if we want to address this in the initial commit or not but: what are the cardinalities we have in mind? Eg, suppose we want to run N validating/mutating webhooks. How many CertRotators does that map to? 1, N, or ...? Also how many webhook controller/reconciler processes?

Comment (maxsmythe) [reply pkg/rotator/rotator.go:51]: The original idea is that all pods that participate in a webhook are running the cert-controller code.

This creates a worst-case scenario of last-write wins, and has the advantage that the pods don't need to coordinate with each other directly and no external management process is needed.

In practice, I think the granularity is at the secret level:

* All pods that are part of the informal "last write wins" ring run the controller against the secret.
* That secret populates all registered webhooks in a single VWH config (not strictly necessary, but simpler than trying to carve out individual webhooks)
* A single secret can fan out to as many VWH configs as we like. Different pods can fan out to different VWH configs.

File: pkg/rotator/rotator.go (hunk 1 above)
Comment (adrianludwin) [reply pkg/rotator/rotator.go:90]: This is just what's checked in now; I don't know the code well myself. My personal preference would be to leave the first checkin as-is (I would have been happy even if it didn't compile for history purposes) and then file issues for all these concerns. wdyt?

**Comment by maxsmythe** [comment]
@adrianludwin Sorry, I should not have merged this. I concur with ekitson's request to add a link in the PR description to the original code.

Also, this commit needs DCO... it looks like it wasn't set to be required as it is in Gatekeeper's repo.

I did a hard reset as I had just merged and wanted to remove the non-DCO commit from the git history and allow for updating the PR descriptions. Could you re-open this PR?

**Comment by adrianludwin** [comment]
I'll leave this one closed, the new PR is #4 .

**Pull request merged**